### PR TITLE
cmake: force linkage of shared libraries

### PR DESCRIPTION
**Pull request by veprbl**
This is brought by an issue that plagued pid_lut builds inside the container:
https://eicweb.phy.anl.gov/EIC/benchmarks/physics_benchmarks/-/jobs/3070278#L1611

Our build process produces two shared objects:
 - A foo_plugin that incorporates src/foo/foo.cc (containing JANA2-related enty point)
 - A foo_library that contains all other *.cc files In order to have a complete set of symbols we ask linker to kindly link foo_library to foo_plugin. Unfortunately, it can ignore our request if it deems that the foo_library symbols are not referenced anywhere. This commit modifies the linking process to incorporate --no-as-needed option. The fancy feature from the
https://gitlab.kitware.com/cmake/cmake/-/commit/42965799b4747ab1e0afa6546be13444f68c1987 makes it so that other libraries (e.g. external ones like podio) can still be linked lazily.

At this point we can remove WITH_STATIC_LIBRARY per-plugin configuration.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -65,6 +65,12 @@ endif()
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
 
+include(CheckLinkerFlag)
+check_linker_flag(CXX "-Wl,--no-as-needed" CXX_LINKER_HAS_no_as_needed)
```
[the remark this answers is not in the document]
Comment (veprbl) [reply CMakeLists.txt:69]:
```suggestion
check_linker_flag(CXX "LINKER:--no-as-needed" CXX_LINKER_HAS_no_as_needed)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -65,6 +65,12 @@ endif()
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
 
+include(CheckLinkerFlag)
```
[the remark this answers is not in the document]
Comment (veprbl) [reply CMakeLists.txt:68]:
```suggestion
# Infrastructure to support use of --no-as-needed
include(CheckLinkerFlag)
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -26,7 +26,7 @@
 #
 # cmake-format: on
 
-cmake_minimum_required(VERSION 3.18)
+cmake_minimum_required(VERSION 3.24)
```
[the remark this answers is not in the document]
Comment (veprbl) [reply CMakeLists.txt:29]: People who are affected will be notified at the time they invoke cmake.

[the remark this answers is not in the document]
Comment (wdconinc) [reply CMakeLists.txt:29]: Yeah, it's mostly for whoever updates the eicrecon spack package on eic-spack, i.e. you or I :-)

Having to chase changed dependency versions is among the most annoying parts of a package maintainer job...

[the remark this answers is not in the document]
Comment (veprbl) [reply CMakeLists.txt:29]: git blame works quite well. I don't see how slapping a release note can make or break this. To increase the requirement - it's either a bad idea or not.

[the remark this answers is not in the document]
Comment (wdconinc) [reply CMakeLists.txt:29]: I fully agree with increasing the requirement to cmake 3.24 :-) It allows us to use some other features too (header sets in particular).

**Comment by veprbl** [comment]
The reco_test issue is to be addressed in #1433. Let's see if physics_benchmarks work now.

**Pull request merged**